lib/src/middleware.dart

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'client.dart';
+import 'handler_client.dart';
+import 'request.dart';
+import 'response.dart';
+
+typedef Client Middleware(Client inner);
+
+typedef Future<Response> ResponseHandler(Response response);
+typedef Future<Request> RequestHandler(Request request);
+
+Middleware createMiddleware({
+  RequestHandler requestHandler,
+  ResponseHandler responseHandler,
+  CloseHandler closeHandler,
+}) {
+  requestHandler ??= (request) async => request;
+  responseHandler ??= (response) async => response;
+  closeHandler ??= () {};
+
+  return (Client inner) {
+    return new HandlerClient(
+          (Request request) async {
+        request = await requestHandler(request);
+
+        final response = await inner.send(request);
+
+        return await responseHandler(response);
+      },
+      () {
+        closeHandler();
+
+        inner.close();
+      },
```

Comment:
> It would be but this ensures that close goes down through the pipeline.

I'm not sure what you mean by "goes down the pipeline".

> As an aside into performance would it be better to just have
> 
> ```
> Future<Request> _defaultRequest(Request request) async => request;
> Future<Response> _defaultResponse(Response response) async => response;
> ```
> etc

Nah, it would be trivial for implementations to hoist these if it's actually relevant to performance.